### ISSUE TITLE
[1706.04599v2] On Calibration of Modern Neural Networks

### ISSUE THREAD

**Issue by riron1206** [riron1206/read_paper#60]
## 概要
https://arxiv.org/abs/1706.04599v2
Temperature scaling
ニューラルネットの予測の確率（確信度）を補正する後処理手法（Softmax関数の入力を温度パラメータT(>0)で割るだけ）

近年のニューラルネットワークは、自信過剰の確率出力する傾向があり、Temperature scalingはそれを修正する後処理方法

Overconfidenceとは、分類タスクにおいてモデルの予測が正しくないにもかかわらず、大きな値を出力してしまう問題である。
これはモデルの出力がそのまま確率的な意味として捉えることはできないことを意味している。
例えば、犬と猫を分類するタスクを考える。
犬の確率が0.9で猫の確率が0.1であるとモデルが予測をした場合に、入力が90%の確率で犬であるとは言えないという問題である。
分類モデルの実世界での応用を考えると、単にモデルが正しい予測をするだけでなく、どんなときに予測を誤ってしまうかを予測することは重要であるが、深層学習のモデルは上記の問題があるとし、出力の値を確率的な意味として考えることはできないとしている。

LenetはOverconfidenceしてないがResnetはOverconfidenceしている例
![image](https://user-images.githubusercontent.com/22429441/109756637-a6def680-7c2b-11eb-848f-6252e4a8101a.png)

## 新規性・差分
**Softmax関数の入力を温度パラメータT(>0)で割るだけで確信度キャリブレーションできること**

![image](https://user-images.githubusercontent.com/22429441/109757019-52884680-7c2c-11eb-9a41-9133aff40e9d.png)

【Temperature Scalingの性質】
- 予測クラスは変わらない
- T=1ならば通常のSoftmax関数と同じ
- Tを大きくしていくと、全てのクラスの予測は同じ値に近づいていく
- TはNLL（負の対数尤度）を最小化するように選択し、検証セットで学習

## 手法
画像とドキュメントの分類データセットを使用して、最先端のアーキテクチャでさまざまな後処理キャリブレーション方法のパフォーマンスを評価

キャリブレーションの評価はExpected calibration error (ECE)
![image](https://user-images.githubusercontent.com/22429441/109771882-b0bf2480-7c40-11eb-9a3d-17c3d080fc3b.png)

確率区間[0, 1]を等間隔のビンの下で離散化し、各予測確率をそれを包含するビンに割り当てる
calibration errorとは、ビン内の予測値が正しい割合（精度）とビン内の確率の平均値（信頼度）との差
Expected calibration errorはビン間の平均値
ラベル yn が K 個の値のうち 1 個の値をとる N 個の入出力ペア（xn, yn）のデータセットについて、ECE は加重平均を計算
Bはビン数、n_bはビンbの予測数、acc(b)とconf(b)はそれぞれビンbの精度と信頼度

## 結果
**ニューラルネットのdepth, width, weight decay, Batch Normalization がOverconfidenceを与えることがわかった**

## コメント
kaggleでためしたnotebook:
- https://www.kaggle.com/anonamename/temperature-scaling # pytorch

公式github:
- https://github.com/gpleiss/temperature_scaling

非公式github:
- https://github.com/cpark321/uncertainty-deep-learning/blob/master/04.%20Temperature_scaling.ipynb  # わかりやすい

参考サイト:
- https://qiita.com/sho4_forever/items/460f124c20298faa07dc
- https://nryotaro.dev/posts/on_calibration_of_modern_neural_networks/
